xds/internal/resolver/cluster_specifier_plugin_test.go

```diff
@@ -20,349 +20,466 @@ package resolver
 	ArbitraryField string `json:"arbitrary_field"`
 }
 
-// TestXDSResolverClusterSpecifierPlugin tests that cluster specifier plugins
-// produce the correct service config, and that the config selector routes to a
-// cluster specifier plugin supported by this service config (i.e. prefixed with
-// a cluster specifier plugin prefix).
-func (s) TestXDSResolverClusterSpecifierPlugin(t *testing.T) {
-	xdsR, xdsC, tcc, cancel := testSetup(t, setupOpts{target: target})
-	defer xdsR.Close()
-	defer cancel()
+func (cspBalancerBuilder) ParseConfig(lbCfg json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+	cfg := &cspBalancerConfig{}
+	if err := json.Unmarshal(lbCfg, cfg); err != nil {
+		return nil, err
+	}
+	return cfg, nil
+
+}
+
+// testClusterSpecifierPlugin is a test cluster specifier plugin which returns
+// an LB policy configuration specifying the cspBalancer.
+type testClusterSpecifierPlugin struct {
+}
+
+func (testClusterSpecifierPlugin) TypeURLs() []string {
+	// The config for this plugin contains a wrapperspb.StringValue, and since
+	// we marshal that proto as an Any proto, the type URL on the latter gets
+	// set to "type.googleapis.com/google.protobuf.StringValue". If we wanted a
+	// more descriptive type URL for this test plugin, we would have to define a
+	// proto package with a message for the configuration. That would be
+	// overkill for a test. Therefore, this seems to be an acceptable tradeoff.
+	return []string{"type.googleapis.com/google.protobuf.StringValue"}
+}
 
+func (testClusterSpecifierPlugin) ParseClusterSpecifierConfig(cfg proto.Message) (clusterspecifier.BalancerConfig, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("testClusterSpecifierPlugin: nil configuration message provided")
+	}
+	any, ok := cfg.(*anypb.Any)
+	if !ok {
+		return nil, fmt.Errorf("testClusterSpecifierPlugin: error parsing config %v: unknown type %T", cfg, cfg)
+	}
+	lbCfg := new(wrapperspb.StringValue)
+	if err := ptypes.UnmarshalAny(any, lbCfg); err != nil {
+		return nil, fmt.Errorf("testClusterSpecifierPlugin: error parsing config %v: %v", cfg, err)
+	}
+	return []map[string]interface{}{{"csp_experimental": cspBalancerConfig{ArbitraryField: lbCfg.GetValue()}}}, nil
+}
+
+// TestResolverClusterSpecifierPlugin tests the case where a route configuration
+// containing cluster specifier plugins is sent by the management server. The
+// test verifies that the service config output by the resolver contains the LB
+// policy specified by the cluster specifier plugin, and the config selector
+// returns the cluster associated with the cluster specifier plugin.
+//
+// The test also verifies that a change in the cluster specifier plugin config
+// result in appropriate change in the service config pushed by the resolver.
+func (s) TestResolverClusterSpecifierPlugin(t *testing.T) {
+	// Env var GRPC_EXPERIMENTAL_XDS_RLS_LB controls whether the xDS client
+	// allows routes with cluster specifier plugin as their route action.
```

Comment:
Nit: do we need this explanation when the env var docstring gives an explanation?

Reply:
Thinking back, I think I added this because the env var name `envconfigXDSRLS` does not describe anything about cluster specifier plugins. Yes, the docstring on `envconfig.XDSRLS` does say the following:
```
	// XDSRLS indicates whether processing of Cluster Specifier plugins and
	// support for the RLS CLuster Specifier is enabled, which can be enabled by
```
But without a comment here (since the variable name is not descriptive enough), the reader would be forced to navigate to the definition of the variable to understand why some seemingly RLS related env variable is set for a test which isn't doing anything specific to RLS.

And taking advantage of your other comment, I'm going to leave it as it.